File: rb/GuildWars2_sdk.rb
# GuildWars2 SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'GuildWars2_types'


class GuildWars2SDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = GuildWars2Utility.new
    @_utility = utility

    config = GuildWars2Config.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = GuildWars2Helpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = GuildWars2Helpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, GuildWars2Features.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    GuildWars2Utility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = GuildWars2Helpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = GuildWars2Helpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = GuildWars2Helpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = GuildWars2Spec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue GuildWars2Error => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = GuildWars2Helpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = GuildWars2Helpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.Achievement.list / client.Achievement.load({ "id" => ... })
  def Achievement(data = nil)
    require_relative 'entity/achievement_entity'
    AchievementEntity.new(self, data)
  end


  # Canonical facade: client.Authenticated.list / client.Authenticated.load({ "id" => ... })
  def Authenticated(data = nil)
    require_relative 'entity/authenticated_entity'
    AuthenticatedEntity.new(self, data)
  end


  # Canonical facade: client.DailyReward.list / client.DailyReward.load({ "id" => ... })
  def DailyReward(data = nil)
    require_relative 'entity/daily_reward_entity'
    DailyRewardEntity.new(self, data)
  end


  # Canonical facade: client.GameMechanic.list / client.GameMechanic.load({ "id" => ... })
  def GameMechanic(data = nil)
    require_relative 'entity/game_mechanic_entity'
    GameMechanicEntity.new(self, data)
  end


  # Canonical facade: client.Guild.list / client.Guild.load({ "id" => ... })
  def Guild(data = nil)
    require_relative 'entity/guild_entity'
    GuildEntity.new(self, data)
  end


  # Canonical facade: client.GuildAuthenticated.list / client.GuildAuthenticated.load({ "id" => ... })
  def GuildAuthenticated(data = nil)
    require_relative 'entity/guild_authenticated_entity'
    GuildAuthenticatedEntity.new(self, data)
  end


  # Canonical facade: client.HomeInstance.list / client.HomeInstance.load({ "id" => ... })
  def HomeInstance(data = nil)
    require_relative 'entity/home_instance_entity'
    HomeInstanceEntity.new(self, data)
  end


  # Canonical facade: client.Item.list / client.Item.load({ "id" => ... })
  def Item(data = nil)
    require_relative 'entity/item_entity'
    ItemEntity.new(self, data)
  end


  # Canonical facade: client.Map.list / client.Map.load({ "id" => ... })
  def Map(data = nil)
    require_relative 'entity/map_entity'
    MapEntity.new(self, data)
  end


  # Canonical facade: client.MapInformation.list / client.MapInformation.load({ "id" => ... })
  def MapInformation(data = nil)
    require_relative 'entity/map_information_entity'
    MapInformationEntity.new(self, data)
  end


  # Canonical facade: client.Miscellaneous.list / client.Miscellaneous.load({ "id" => ... })
  def Miscellaneous(data = nil)
    require_relative 'entity/miscellaneous_entity'
    MiscellaneousEntity.new(self, data)
  end


  # Canonical facade: client.Story.list / client.Story.load({ "id" => ... })
  def Story(data = nil)
    require_relative 'entity/story_entity'
    StoryEntity.new(self, data)
  end


  # Canonical facade: client.StructuredPvP.list / client.StructuredPvP.load({ "id" => ... })
  def StructuredPvP(data = nil)
    require_relative 'entity/structured_pv_p_entity'
    StructuredPvPEntity.new(self, data)
  end


  # Canonical facade: client.TradingPost.list / client.TradingPost.load({ "id" => ... })
  def TradingPost(data = nil)
    require_relative 'entity/trading_post_entity'
    TradingPostEntity.new(self, data)
  end


  # Canonical facade: client.WorldVsWorld.list / client.WorldVsWorld.load({ "id" => ... })
  def WorldVsWorld(data = nil)
    require_relative 'entity/world_vs_world_entity'
    WorldVsWorldEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = GuildWars2SDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/guildwars2_sdk.py
# GuildWars2 SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import GuildWars2Utility
from core.spec import GuildWars2Spec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import GuildWars2BaseFeature
from features import _make_feature


class GuildWars2SDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = GuildWars2Utility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return GuildWars2Utility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = GuildWars2Spec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def achievement(self):
        """Idiomatic facade: client.achievement.list() / client.achievement.load({"id": ...})."""
        from entity.achievement_entity import AchievementEntity
        cached = getattr(self, "_achievement", None)
        if cached is None:
            cached = AchievementEntity(self, None)
            self._achievement = cached
        return cached

    def Achievement(self, data=None):
        # Deprecated: use client.achievement instead.
        from entity.achievement_entity import AchievementEntity
        return AchievementEntity(self, data)


    @property
    def authenticated(self):
        """Idiomatic facade: client.authenticated.list() / client.authenticated.load({"id": ...})."""
        from entity.authenticated_entity import AuthenticatedEntity
        cached = getattr(self, "_authenticated", None)
        if cached is None:
            cached = AuthenticatedEntity(self, None)
            self._authenticated = cached
        return cached

    def Authenticated(self, data=None):
        # Deprecated: use client.authenticated instead.
        from entity.authenticated_entity import AuthenticatedEntity
        return AuthenticatedEntity(self, data)


    @property
    def daily_reward(self):
        """Idiomatic facade: client.daily_reward.list() / client.daily_reward.load({"id": ...})."""
        from entity.daily_reward_entity import DailyRewardEntity
        cached = getattr(self, "_daily_reward", None)
        if cached is None:
            cached = DailyRewardEntity(self, None)
            self._daily_reward = cached
        return cached

    def DailyReward(self, data=None):
        # Deprecated: use client.daily_reward instead.
        from entity.daily_reward_entity import DailyRewardEntity
        return DailyRewardEntity(self, data)


    @property
    def game_mechanic(self):
        """Idiomatic facade: client.game_mechanic.list() / client.game_mechanic.load({"id": ...})."""
        from entity.game_mechanic_entity import GameMechanicEntity
        cached = getattr(self, "_game_mechanic", None)
        if cached is None:
            cached = GameMechanicEntity(self, None)
            self._game_mechanic = cached
        return cached

    def GameMechanic(self, data=None):
        # Deprecated: use client.game_mechanic instead.
        from entity.game_mechanic_entity import GameMechanicEntity
        return GameMechanicEntity(self, data)


    @property
    def guild(self):
        """Idiomatic facade: client.guild.list() / client.guild.load({"id": ...})."""
        from entity.guild_entity import GuildEntity
        cached = getattr(self, "_guild", None)
        if cached is None:
            cached = GuildEntity(self, None)
            self._guild = cached
        return cached

    def Guild(self, data=None):
        # Deprecated: use client.guild instead.
        from entity.guild_entity import GuildEntity
        return GuildEntity(self, data)


    @property
    def guild_authenticated(self):
        """Idiomatic facade: client.guild_authenticated.list() / client.guild_authenticated.load({"id": ...})."""
        from entity.guild_authenticated_entity import GuildAuthenticatedEntity
        cached = getattr(self, "_guild_authenticated", None)
        if cached is None:
            cached = GuildAuthenticatedEntity(self, None)
            self._guild_authenticated = cached
        return cached

    def GuildAuthenticated(self, data=None):
        # Deprecated: use client.guild_authenticated instead.
        from entity.guild_authenticated_entity import GuildAuthenticatedEntity
        return GuildAuthenticatedEntity(self, data)


    @property
    def home_instance(self):
        """Idiomatic facade: client.home_instance.list() / client.home_instance.load({"id": ...})."""
        from entity.home_instance_entity import HomeInstanceEntity
        cached = getattr(self, "_home_instance", None)
        if cached is None:
            cached = HomeInstanceEntity(self, None)
            self._home_instance = cached
        return cached

    def HomeInstance(self, data=None):
        # Deprecated: use client.home_instance instead.
        from entity.home_instance_entity import HomeInstanceEntity
        return HomeInstanceEntity(self, data)


    @property
    def item(self):
        """Idiomatic facade: client.item.list() / client.item.load({"id": ...})."""
        from entity.item_entity import ItemEntity
        cached = getattr(self, "_item", None)
        if cached is None:
            cached = ItemEntity(self, None)
            self._item = cached
        return cached

    def Item(self, data=None):
        # Deprecated: use client.item instead.
        from entity.item_entity import ItemEntity
        return ItemEntity(self, data)


    @property
    def map(self):
        """Idiomatic facade: client.map.list() / client.map.load({"id": ...})."""
        from entity.map_entity import MapEntity
        cached = getattr(self, "_map", None)
        if cached is None:
            cached = MapEntity(self, None)
            self._map = cached
        return cached

    def Map(self, data=None):
        # Deprecated: use client.map instead.
        from entity.map_entity import MapEntity
        return MapEntity(self, data)


    @property
    def map_information(self):
        """Idiomatic facade: client.map_information.list() / client.map_information.load({"id": ...})."""
        from entity.map_information_entity import MapInformationEntity
        cached = getattr(self, "_map_information", None)
        if cached is None:
            cached = MapInformationEntity(self, None)
            self._map_information = cached
        return cached

    def MapInformation(self, data=None):
        # Deprecated: use client.map_information instead.
        from entity.map_information_entity import MapInformationEntity
        return MapInformationEntity(self, data)


    @property
    def miscellaneous(self):
        """Idiomatic facade: client.miscellaneous.list() / client.miscellaneous.load({"id": ...})."""
        from entity.miscellaneous_entity import MiscellaneousEntity
        cached = getattr(self, "_miscellaneous", None)
        if cached is None:
            cached = MiscellaneousEntity(self, None)
            self._miscellaneous = cached
        return cached

    def Miscellaneous(self, data=None):
        # Deprecated: use client.miscellaneous instead.
        from entity.miscellaneous_entity import MiscellaneousEntity
        return MiscellaneousEntity(self, data)


    @property
    def story(self):
        """Idiomatic facade: client.story.list() / client.story.load({"id": ...})."""
        from entity.story_entity import StoryEntity
        cached = getattr(self, "_story", None)
        if cached is None:
            cached = StoryEntity(self, None)
            self._story = cached
        return cached

    def Story(self, data=None):
        # Deprecated: use client.story instead.
        from entity.story_entity import StoryEntity
        return StoryEntity(self, data)


    @property
    def structured_pv_p(self):
        """Idiomatic facade: client.structured_pv_p.list() / client.structured_pv_p.load({"id": ...})."""
        from entity.structured_pv_p_entity import StructuredPvPEntity
        cached = getattr(self, "_structured_pv_p", None)
        if cached is None:
            cached = StructuredPvPEntity(self, None)
            self._structured_pv_p = cached
        return cached

    def StructuredPvP(self, data=None):
        # Deprecated: use client.structured_pv_p instead.
        from entity.structured_pv_p_entity import StructuredPvPEntity
        return StructuredPvPEntity(self, data)


    @property
    def trading_post(self):
        """Idiomatic facade: client.trading_post.list() / client.trading_post.load({"id": ...})."""
        from entity.trading_post_entity import TradingPostEntity
        cached = getattr(self, "_trading_post", None)
        if cached is None:
            cached = TradingPostEntity(self, None)
            self._trading_post = cached
        return cached

    def TradingPost(self, data=None):
        # Deprecated: use client.trading_post instead.
        from entity.trading_post_entity import TradingPostEntity
        return TradingPostEntity(self, data)


    @property
    def world_vs_world(self):
        """Idiomatic facade: client.world_vs_world.list() / client.world_vs_world.load({"id": ...})."""
        from entity.world_vs_world_entity import WorldVsWorldEntity
        cached = getattr(self, "_world_vs_world", None)
        if cached is None:
            cached = WorldVsWorldEntity(self, None)
            self._world_vs_world = cached
        return cached

    def WorldVsWorld(self, data=None):
        # Deprecated: use client.world_vs_world instead.
        from entity.world_vs_world_entity import WorldVsWorldEntity
        return WorldVsWorldEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
